File: 리트코드/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
        public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> used = new HashMap<>();
        int maxLen = 0;
        int left = 0, right = 0;

        for (char c : s.toCharArray()) {

            if (used.containsKey(c) && left <= used.get(c)) {
                // 마지막으로 저장되었던 위치 다음으로 이동
                left = used.get(c) + 1;
            }
            
            maxLen = Math.max(maxLen, right - left + 1);
            used.put(c,right);
            right++;
            
        }
        return maxLen;
    }

}

File: 리트코드/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var used :HashMap<Char,Int> = HashMap()
        var left = 0
        var right = 0
        var maxLength = 0

        for (c in s.toCharArray()) {
            if (left <= used.getOrDefault(c, -1)) {
                left = used.get(c)!! + 1
            } else {
                maxLength = Math.max(maxLength,right - left + 1)
            }
            used.put(c,right)
            right++
        }
        return maxLength
    }
}